File: hydra-java/src/main/java/hydra/dsl/prims/Logic.java
package hydra.dsl.prims;

import hydra.core.Term;
import hydra.lib.logic.And;
import hydra.lib.logic.IfElse;
import hydra.lib.logic.Not;
import hydra.lib.logic.Or;

/**
 * DSL interface providing logic primitive operations.
 */
public interface Logic {
    /**
     * Returns a term representing the and primitive operation.
     * Logical conjunction.
     *
     * @return a term for logical and
     */
    static Term and() {
        return new And().term();
    }

    /**
     * Returns a term representing the ifElse primitive operation.
     * Conditional expression.
     *
     * @return a term for conditional expressions
     */
    static Term ifElse() {
        return new IfElse().term();
    }

    /**
     * Returns a term representing the not primitive operation.
     * Logical negation.
     *
     * @return a term for logical not
     */
    static Term not() {
        return new Not().term();
    }

    /**
     * Returns a term representing the or primitive operation.
     * Logical disjunction.
     *
     * @return a term for logical or
     */
    static Term or() {
        return new Or().term();
    }
}

File: hydra-clojure/src/main/clojure/hydra/lib/logic.clj
(ns hydra.lib.logic)

;; and :: Bool -> Bool -> Bool
(def hydra_lib_logic_and
  (fn [a] (fn [b] (and a b))))

;; if_else :: Bool -> a -> a -> a
(def hydra_lib_logic_if_else
  (fn [cond_] (fn [then_] (fn [else_] (if cond_ then_ else_)))))

;; not :: Bool -> Bool
(def hydra_lib_logic_not
  (fn [a] (not a)))

;; or :: Bool -> Bool -> Bool
(def hydra_lib_logic_or
  (fn [a] (fn [b] (or a b))))
